### ISSUE TITLE
CCT invalid signature error

### ISSUE THREAD

**Issue by tobischo** [railslove/epics#103]
We have been using epics for years to download files through C52 and C53.

For a slightly different use case we wanted to use epics now to pass in pain.001.001.03 files (generated through sepa_king) through the CCT action.

However we keep getting an invalid signature error (error code 28 at FIDUCIA/ATRUVIA).

That means we managed to successfully upload the file, so the general authentication worked.
The file could be correctly decrypted, however it lacked the corresponding signature (I suppose A006)

In the readme a few banks are listed where epics works correctly.
It does not however list which order types are executed there.
Now I am wondering whether for the cases with the banks where it worked, the order types CCT, CDD, or CDB where amongst those that are executed.

Do you happen to have any more information on that?

I would love to know whether I am just doing something wrong or whether epics actually has an issue here.

**Comment by bumi** [comment]
@tobischo thanks for opening this issue. 
Sadly I am not using (or working with) epics right now and I don't fully remember what banks and calls I have been involved with. But I am pretty sure that CCT and CDD were among them. 

Sadly I can not be of any help here right now. sorry.

**Comment by tobischo** [comment]
@bumi 
We now got an EBICS Dummy server from subsembly to test exactly this.
I can at the very least already tell you that this server also does not like the order data signature generation.
So it stands to reason that, as it is right now, the signature generation here does not work correctly in epics.

**Comment by bumi** [comment]
ah interesting, do you know which banks use subsembly? I think they are always a bit different. Do you have an error message? Maybe @Uepsilon or @kangguru have ideas.
Would be great to have a PR here if you find out what the problem is. 

A maintenance update with some updated dependencies of epics is actually also needed.

**Comment by tobischo** [comment]
The error message from FIDUCIA/ATRUVIA was basically none at all - upload worked but the processing failed later.

The Subsembly EBICS Dummy Server upload resulted in the following being raised
```
Epics::Error::BusinessError (EBICS_SIGNATURE_VERIFICATION_FAILED - Verification of the ES has failed In the case of asynchronouslyimplemented orders, the error can occur during preliminary verification.)
```
The nice thing about this Dummy Server is that it allows for immediate feedback, which is what I didn't have with the FIDUCIA/ATRUVIA live setup.

Now I actually have made a tiny bit of progress so far, but more out of trying something "random" out of pure curiosity:

The key size we chose was 4096, as the BSI recommends at least 3000bit for keys that should live longer than 2022.

With the Dummy Server the upload passes when I chose a 2048 bit key (default size), but it fails when I chose 3072 or 4096.
So it stands to reason that the current implementation is not able to deal with that part, but we could probably get it working productively if we decided to reduce the key size.

**Comment by tobischo** [comment]
For the sake of trying to figure out where exactly the problem lies I replaced the `key.rb#sign` method with this:
```ruby
require 'open3'
require 'tempfile'

def sign(msg)
  file = Tempfile.new('foo')
  file.write(key.to_pem)
  file.close

  command = "echo '#{Base64.encode64(msg)}' | base64 -d | openssl dgst -sha256 -sign #{file.path} -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:32 | base64 -w0"

  stdout, stderr, status = Open3.capture3(command)

  file.unlink

  return stdout
end
```
which in the end is the exact command that is suggested in `key_spec.rb` for verifying that it works.

With this I also managed to get it working for keys with the size 3072bit and 4096bit.

So the issue appears to be definitely somewhere inside of `sign` or the methods called from there.

**Comment by tobischo** [comment]
The equivalent of that in ruby should be:
```ruby
  def sign(msg)
    return Base64.encode64(
             key.sign_pss(
               'SHA256',
               msg,
               salt_length: :digest,
               mgf1_hash:   'SHA256',
             )
           ).gsub("\n", '')
end
```

At least that is also working against the subsembly dummy server.
I will verify this, probably tomorrow, against the FIDUCIA/ATRUVIA system.

Should this work my proposal would be to replace the current implementation of `sign` with this version.
That should also allow removing the implementation of MGF1.

You probably have more context than I do to check that this is working.

**Comment by tobischo** [comment]
I managed to verify it today and actually execute transfers with our regular account and 4096bit keys.

I will prepare a pull request with the adjustments I think are necessary in the next few days

**Comment by bumi** [comment]
I am happy to hear that. Awesome @tobischo! 
Such an interesting problem, thanks for debugging and solving this!

**Comment by tobischo** [comment]
> and ebics is also not necessarily something that you do in your private-side projects :D

Which is probably caused by banks not providing EBICS access in most cases unless you have a business account with them.

Considering that we bought the license for the subsembly EBICS test server for a year now, I considered porting this to go or adding EBICS 3.0 support.

However that will also require a lot of spare time or time at work, so chances are low on that

**Comment by tobischo** [comment]
As #108 and #105 are still pending, we released a version with the fix from #105 onto our own internal gem server

**Comment by tobischo** [comment]
I am going to close this issue, as I already merged the PR for it.
There are still a bunch of other things that could be added to a new release
